tests/integration/e2e/conftest.py

```diff
@@ -162,23 +163,106 @@ async def _deploy_client(role: Literal["producer", "consumer"]):
             num_units=1,
             series="jammy",
             channel="edge",
-            config={"topic-name": "HOT-TOPIC", "extra-user-roles": role},
+            config=config,
         )
         await ops_test.model.wait_for_idle(apps=[generated_app_name])
 
+        return generated_app_name
+
+    logger.info(f"setting up data_integrator - current apps {apps}")
+    yield _deploy_data_integrator
+
+    logger.info(f"tearing down {apps}")
+    for app in apps:
+        logger.info(f"tearing down {app}")
+        await ops_test.model.applications[app].remove()
+
+    await ops_test.model.wait_for_idle(apps=[kafka], idle_period=30, status="active", timeout=1800)
+
+
+@pytest.fixture(scope="function")
+async def deploy_test_app(ops_test: OpsTest, kafka, tls):
+    """Factory fixture for deploying + tearing down client applications."""
+    # tracks deployed app names for teardown later
+    apps = []
+
+    async def _deploy_test_app(
+        role: Literal["producer", "consumer"],
+        topic_name: str = "test-topic",
+        consumer_group_prefix: Optional[str] = None,
+        num_messages: int = 1500,
+    ):
+        """Deploys client with specified role and uuid."""
+        if not ops_test.model:  # avoids a multitude of linting errors
+            raise RuntimeError("model not set")
+
+        # uuid to avoid name clashes for same applications
+        key = "".join(random.choices(string.ascii_lowercase, k=4))
+        generated_app_name = f"{role}-{key}"
+        apps.append(generated_app_name)
+
+        logger.info(f"{generated_app_name=} - {apps=}")
+
+        config = {"role": role, "topic_name": topic_name, "num_messages": num_messages}
+
+        if consumer_group_prefix:
+            config["consumer_group_prefix"] = consumer_group_prefix
+
+        # todo substitute with the published charm
+        await ops_test.model.deploy(
+            KAFKA_TEST_APP_CHARM_NAME,
+            application_name=generated_app_name,
+            num_units=1,
+            series="jammy",
+            channel="edge",
+            config=config,
+        )
+        await ops_test.model.wait_for_idle(
+            apps=[generated_app_name], idle_period=20, status="active"
+        )
+
+        # Relate with TLS operator
+        if tls:
+            await ops_test.model.add_relation(generated_app_name, TLS_CHARM_NAME)
+            await ops_test.model.wait_for_idle(
+                apps=[generated_app_name, TLS_CHARM_NAME],
+                idle_period=30,
+                status="active",
+                timeout=1800,
+            )
+
+        # Relate with MongoDB
+        await ops_test.model.add_relation(generated_app_name, DATABASE_CHARM_NAME)
```

Comment:
**thought*** If easy to do, I would allow also for MongoDB to be fed in an external model. In this case, we should not tear the existing MongoDB down, such that we can do analytics on top to understand what happened during the test. 

If this is more complicated and not done within 30min-1h, we can postpone this to a separate ticket. In this case please, just create that in the backlog though and link it here